### PR TITLE
Adds a default-implemented `SecondaryMap::remove`

### PR DESCRIPTION
**Pull request by aborgna-q**

### CODE REVIEW

**Comment by aborgna-q** [comment]
- `must_use` is a rust flag for methods where not checking the returned value is suspicious (e.g. methods that return a Result). This has nothing to do with relevant types, one can always explicitly ignore the results with `let _ = method()`.

- Yes, I'm not sure about it either. `take` is a standard name for "retrieve and leave a default", so I wouldn't rename. But there are instances where we want to just drop the value, and `remove` would be similar with the rest of the API;
```rust
graph.remove_node(node);
hierarchy.remove(node);
secondary_map.???(node);
```

**Comment by acl-cqc** [comment]
> * `must_use` is a rust flag for methods where not checking the returned value is suspicious (e.g. methods that return a Result). This has nothing to do with relevant types, one can always explicitly ignore the results with `let _ = method()`.
> * Yes, I'm not sure about it either. `take` is a standard name for "retrieve and leave a default", so I wouldn't rename. But there are instances where we want to just drop the value, and `remove` would be similar with the rest of the API;
> 
> ```rust
> graph.remove_node(node);
> hierarchy.remove(node);
> secondary_map.???(node);
> ```

Hmmmm, your second set of examples is quite convincing, I agree. So maybe `HashMap` is the odd one out (`remove` returning a value) ! Indeed, perhaps HashSet is a better example (in agreement with your 3) - here `take` is much as you say and `remove` returns a bool (true = mutation happened). How about that?

**Comment by acl-cqc** [comment]
I realize that's a bit harder to implement as `remove` is no longer strictly a function of `take`. Hmmm, so in fact there are two possibilities. Suppose we first do `m.put(key, default())`, then
1. `m.remove(key) == true` - in which case `remove` cannot be a function of take as it must inspect whether the entry was there - and explicitly storing the default is different from not storing anything, with *this difference revealed only by remove*. The italics sounds a bit nasty, so 
2. `m.remove(key) == false`, which is to say, the original `m.put(key, default())` was a no-op that didn't mutate the map. In this case the trait can define `remove` as `take() != default()`. This is neater in many ways but relies on a strong notion of equality.

I could go either way there!!

**Comment by aborgna-q** [comment]
Yes, that's the main problem with that.
SecondaryMaps are not required to carry information about what has been allocated, only to return the right values. It's perfectly valid to do something like this to preserve memory.
```rust
if node_val != default {
  secondary_map.set(node, node_val);
}
```
(Note: this is automatically done by some impls)

Returning a flag for whether the element was present when deleting it goes kind of against it. For the same reason we do not have a `contains(key)`.

**Comment by aborgna-q** [comment]
Another problem as you say is that we don't always have `V: Eq`, so distinguishing already-erased elements may be impossible.

Note: I added a drive-by change to `UnmanagedDenseMap::take` to avoid allocating memory on some cases.

File: src/unmanaged.rs [hunk 1]
```diff
@@ -372,7 +372,10 @@ where
     #[inline]
     fn take(&mut self, key: K) -> V {
         let default = self.default.clone();
-        mem::replace(&mut self[key], default)
+        match self.try_get_mut(key) {
+            Some(val) => mem::replace(val, default),
```
Comment (acl-cqc) [inline src/unmanaged.rs:376]: Might be worth a comment that this'll leave trailing `default`s - if those are at the end of the `data` vector, they could be trimmed to save memory, but the only resizing methods truncate (i.e. happily drop non-default values), which makes this hard (albeit not impossible) for the owner/client/caller to do.

**Review by acl-cqc — APPROVED** [review]
Yeah OK. I'm a bit meh but don't see anything better as it doesn't seem possible to be consistent with other classes' `remove` methods :(

**Pull request merged**